src/ArmTemplates/Common/Templates/Abstractions/Template.Generic.cs

```diff
@@ -0,0 +1,41 @@
+﻿// --------------------------------------------------------------------------
+//  <copyright file="Template.Generic.cs" company="Microsoft">
+//      Copyright (c) Microsoft Corporation. All rights reserved.
+//  </copyright>
+// --------------------------------------------------------------------------
+
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Templates.Abstractions
+{
+    public class Template<TTemplateResources> : Template
+        where TTemplateResources : ITemplateResources, new()
+    {
+        /// <summary>
+        /// Specifically typed resources of specific template type.
+        /// Use it for explicit code type usage rather than base <cref="Resources"/> property of base class.
+        /// It will not be serialized
+        /// </summary>
+        [JsonIgnore]
+        public TTemplateResources SpecificResources { get; set; }
+
+        /// <summary>
+        /// Returns TemplateResources collection by building specific resources of current template
+        /// </summary>
+        [JsonProperty(Order = 10)]
+        public new TemplateResource[] Resources => this.SpecificResources.BuildTemplateResources();
```

Comment:
Let's use `override` instead of `new` here and in the method below?

Right now, if you create a `Template<TTemplateResources>` with some `TTemplateResources`, but save it into a `Template`, calling `.HasResources()` will return false and `.Resources` will return null.

Reply:
Great suggestion, didn't think of that solution. Even once I met it in the runtime, so changed it.
Tests are passing :)

Reply:
Yeah. The "new" keyboard is basically just for suppressing the warning that you get if you don't use new nor override, because the behaviour is kinda unexpected. Adding "new" is basically saying "yeah it's odd, but I know what I'm doing" 😄